appserver/web/gf-weld-connector/src/main/java/org/glassfish/weld/connector/WeldUtils.java

```diff
@@ -195,25 +195,49 @@ public static boolean hasCDIEnablingAnnotations(DeploymentContext context, URI p
      * @return true, if there is at least one bean annotated with a qualified annotation in the specified paths
      */
     public static boolean hasCDIEnablingAnnotations(DeploymentContext context, Collection<URI> paths) {
-        List<String> result = new ArrayList<String>();
+        final String METADATA_KEY = "implicitBeanPaths";
+        Set<URI> implicitBeanArchivePaths = (Set<URI>)context.getTransientAppMetadata().get(METADATA_KEY);
+
+        if (implicitBeanArchivePaths == null) {
+            implicitBeanArchivePaths = findImplicitBeanArchivePaths(implicitBeanArchivePaths, context);
+            context.addTransientAppMetaData(METADATA_KEY, implicitBeanArchivePaths);
+        }
+
+        for (URI path : paths) {
+            if (implicitBeanArchivePaths.contains(path)) {
+                return true;
+            }
+        }
+        return false;
+    }
 
+    private static Set<URI> findImplicitBeanArchivePaths(Set<URI> pathsWithImplicitCDIBeans, DeploymentContext context) {
+        pathsWithImplicitCDIBeans = new HashSet<>();
```

Comment:
New, private method - maybe drop `pathsWithImplicitCDIBeans` parameter if the first thing is to overwrite it with empty set?

Reply:
Sure, I’ll change it. It’s a leftover after automatic refactoring.